### PR TITLE
KAFKA-4274: offsetsForTimes() hang on empty map.

### PR DESCRIPTION
**Pull request by becketqin**

### CODE REVIEW

**Comment by hachikuji** [comment]
@becketqin Thanks for the patch. Another thing I noticed is that we don't have a null check for the map passed to `offsetsForTimes`. Do you think it's worth adding one?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java [hunk 1]
```diff
@@ -362,6 +362,9 @@ else if (strategy == OffsetResetStrategy.LATEST)
 
     public Map<TopicPartition, OffsetAndTimestamp> getOffsetsByTimes(Map<TopicPartition, Long> timestampsToSearch,
                                                                      long timeout) {
+        if (timestampsToSearch.isEmpty())
+            return Collections.EMPTY_MAP;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java]: `Collections.emptyMap()` is the right one to use. `EMPTY_MAP` should give you a generic warning.

**Comment by becketqin** [comment]
@hachikuji I am not sure. I checked on the other methods such as `seekToBeginning()`, `resume()`, `pause()`. It seems that null pointer is not checked there either. I do not feel strongly on this because either NPE or IllegalArguementException will be thrown. But maybe we want to have the same behavior for all the interfaces?

**Comment by hachikuji** [comment]
@becketqin We do have null checks in `subscribe` and `assign`, so it's not like there's no precedent (as usual, it's difficult being consistent on these things). Speaking generally, I think an explicit check and an `IllegalArgumentException` with a clear message is a better user experience and a safer practice than depending on the code to raise an NPE. That said, let's not block the patch on this if you'd rather consider it separately.

**Comment by hachikuji** [comment]
LGTM. Merging to trunk and 0.10.1.

**Comment by becketqin** [comment]
@hachikuji Thanks. Yes, I think it is probably clearer to the users to have an IllegalArgumentException with a message. Let's do it in a separate patch for all the interfaces.

**Pull request closed without merging**